src/ulong_extras/profile/p-mulmod_shoup.c

```diff
@@ -0,0 +1,118 @@
+/*
+   Copyright 2024 (C) Vincent Neiger
+
+   This file is part of FLINT.
+
+   FLINT is free software: you can redistribute it and/or modify it under
+   the terms of the GNU Lesser General Public License (LGPL) as published
+   by the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.  See <https://www.gnu.org/licenses/>.
+   */
+
+#include "profiler.h"
+#include "ulong_extras.h"
+
+#define NB_ITER 1000
+
+void sample(void * arg, ulong count)
+{
+    nn_ptr array = (nn_ptr) flint_malloc(NB_ITER*sizeof(ulong));
+    FLINT_TEST_INIT(state);
+
+    for (ulong i = 0; i < count; i++)
+    {
+        ulong bits = n_randint(state, FLINT_BITS - 1) + 1;  // 1...63
+        ulong d = n_randbits(state, bits);  // 0 < d < 2**(FLINT_BITS-1) required by mulmod_shoup
+        ulong a = n_randlimb(state);  // a is arbitrary
+
+        for (ulong j = 0; j < NB_ITER; j++)
+            array[j] = n_randint(state, d);  // must be < d
+
+        prof_start();
+        for (ulong j = 0; j < NB_ITER; j++)
+        {
+            ulong aj, r;
+            udiv_qrnnd(aj, r, array[j], UWORD(0), d);
+            array[j] = n_mulmod_shoup(array[j], a, aj, d);
+        }
+        prof_stop();
+    }
+
+    flint_rand_clear(state);
```

Comment:
`FLINT_TEST_INIT` should be paired with a `FLINT_TEST_CLEANUP`. I don't know if it is documented or not. (As a side-note, perhaps it could be renamed to `FLINT_TEST_CLEAR` to keep the semantics consistent.)

Reply:
Ah, I just did this by copy-pasting from some other p-xxx.c file and was wondering about this but did not check. It would be much cleaner indeed; I will do a separate PR (unless there is only a couple of occurrences).

Reply:
Done, see https://github.com/flintlib/flint/pull/2057